adapter/src/ethereum.rs

```diff
@@ -1,101 +1,492 @@
+                    .query("privileges", verified.from, None, Options::default(), None)
+                    .wait()
+                    .map_err(|_| map_error("failed query priviledge level on contract"))?;
+
+                if priviledge_level == *PRIVILEGE_LEVEL_NONE {
+                    return Err(AdapterError::Authorization(
+                        "insufficient privilege".to_string(),
+                    ));
+                }
+                Session {
+                    era: verified.payload.era,
+                    uid: identity.to_owned(),
+                }
+            }
+            None => Session {
+                era: verified.payload.era,
+                uid: verified.from,
+            },
+        };
+
+        session_tokens.insert(token_id, sess.clone());
+        Ok(sess)
+    }
+
+    fn get_auth(&self, validator: &ValidatorDesc) -> AdapterResult<String> {
+        let mut authorization_tokens = self.authorization_tokens.borrow_mut();
+        match (
+            self.wallet.borrow().clone(),
+            authorization_tokens.get(&validator.id),
+        ) {
+            (Some(_), Some(token)) => Ok(token.to_owned()),
+            (Some(wallet), None) => {
+                let era = Utc::now().timestamp() as f64 / 60000.0;
+                let payload = Payload {
+                    id: validator.id.clone(),
+                    era: era.floor() as i64,
+                    identity: None,
+                    address: None,
+                };
+                let token = ewt_sign(&wallet, &self.keystore_pwd, &payload)
+                    .map_err(|_| map_error("Failed to sign token"))?;
+
+                authorization_tokens.insert(validator.id.clone(), token.clone());
+
+                Ok(token)
+            }
+            (_, _) => Err(AdapterError::Configuration(
+                "failed to unlock wallet".to_string(),
+            )),
+        }
+    }
+}
+
+fn check_validator_id_checksum(channel: &Channel) -> bool {
+    channel
+        .spec
+        .validators
+        .into_iter()
+        .any(|v| v.id != eth_checksum::checksum(&v.id))
+}
+
+fn hash_message(message: &str) -> [u8; 32] {
+    let eth = "\x19Ethereum Signed Message:\n";
+    let message_length = message.len();
+
+    let encoded = format!("{}{}{}", eth, message_length, message);
+
+    let mut result = Keccak::new_keccak256();
+    result.update(&encoded.as_bytes());
+
+    let mut res: [u8; 32] = [0; 32];
+    result.finalize(&mut res);
+
+    res
+}
+
+fn map_error(err: &str) -> AdapterError {
+    AdapterError::Failed(err.to_string())
+}
+
+fn get_contract(
+    config: &Config,
+    contract_address: Address,
+    abi: &[u8],
+) -> Result<Contract<Http>, Box<dyn Error>> {
+    let (_eloop, transport) = web3::transports::Http::new(&config.ethereum_network)?;
+    let web3 = web3::Web3::new(transport);
+    let contract = Contract::from_json(web3.eth(), contract_address, abi)?;
```

Comment:
side note: those 2 lines can be simplified, since you already have a result and you don't need `?`
 and then `Ok(...)`, so this:
```
let contract = Contract::from_json(web3.eth(), contract_address, abi)?;
Ok(contract)
```

becomes:
`Contract::from_json(web3.eth(), contract_address, abi)`

Reply:
This is incorrect 

one is of type
```
Result<_, std::boxed::Box<(dyn std::error::Error + 'static)>>
```
the other is of type
```
 std::result::Result<_, ethabi::errors::Error>
```
? boxes the error

Reply:
Ahh cool, didn't know that, I thought it will automatically do that even without `?`